### PR TITLE
[GPU] implement additional debug features to dump runtime memory pool

### PR DESCRIPTION
**Pull request by ahnyoung-paul**
### Details:
- *Set iterations to dump runtime memory pool*
- *Set csv file path to dump runtime memory pool*
- *Update runtime memory pool dump data*

### Tickets:
 - *138484*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/network.cpp [hunk 1]
```diff
@@ -1241,7 +1244,42 @@ void network::execute_impl(const std::vector<event::ptr>& events) {
...
+            int64_t host_mem_size      = get_engine().get_used_device_memory(allocation_type::usm_host);
+            int64_t device_mem_size    = get_engine().get_used_device_memory(allocation_type::usm_device);
+            int64_t mem_pool_size      = get_memory_pool().get_total_mem_pool_size();
+            int64_t etc_size           = device_mem_size - mem_pool_size - const_mem_size - var_mem_size;
+            GPU_DEBUG_COUT << "Memory statistics for (net_id:" << get_id() << ", iter:" << curr_iter << ") host_mem_size: " << get_mb_size(host_mem_size)
+                    << "MB, device_mem_size: " << get_mb_size(device_mem_size) << "MB (mem_pool_size: " << get_mb_size(mem_pool_size)
+                    << "MB, const_mem_size: " << get_mb_size(const_mem_size) << "MB, var_mem_size: " << get_mb_size(var_mem_size)
+                    << "MB, etc: " << std::fixed << get_mb_size(etc_size) << "MB)" << std::endl;
+            GPU_DEBUG_COUT << "============================================================================" << std::endl;
+        }
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/network.cpp:1282]: what about adding this functionality as a separate function and call the function here? This is a main execution loop and it contains too much code for debugging :(

File: src/plugins/intel_gpu/src/runtime/debug_configuration.cpp [hunk 2]
```diff
@@ -333,6 +339,30 @@ debug_configuration::debug_configuration()
...
+            std::string token;
+            while (ss >> token) {
+                try {
+                    dump_runtime_memory_pool_iters.insert(static_cast<int64_t>(std::stol(token)));
+                } catch(const std::exception& ex) {
+                    dump_runtime_memory_pool_iters.clear();
+                    GPU_DEBUG_COUT << "OV_GPU_DumpRuntimeMemoryPoolIters was ignored. It cannot be parsed to integer array." << std::endl;
+                    break;
+                }
+            }
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/runtime/debug_configuration.cpp:362]: could you refactor this code a bit to have something like `std::vector<int> parse_int_array(const char *string)`?

File: src/plugins/intel_gpu/src/runtime/debug_configuration.cpp [hunk 3]
```diff
@@ -333,6 +339,30 @@ debug_configuration::debug_configuration()
         }
     }
 
+    if (dump_runtime_memory_pool_iters_str.size() > 0) {
+        // eliminate '"' from string to avoid parasing error
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/runtime/debug_configuration.cpp:343]: typo: `parasing` -> `parsing`

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp [hunk 4]
```diff
@@ -309,25 +309,118 @@ void memory_pool::clear_pool_for_network(uint32_t network_id) {
...
             }
+            GPU_DEBUG_COUT << "   - min of the memory pool entry: " << min_percentage << std::endl;
+            GPU_DEBUG_COUT << "   - max of the memory pool entry: " << max_percentage << std::endl;
         }
     }
+    if (total_mem_pool_size > 0.f) {
+        GPU_DEBUG_COUT << "========== padded pool efficiency: "
+            << std::to_string(static_cast<float>(total_mem_request / total_mem_pool_size))
+            << " (Total memory request : " << get_mb_size(total_mem_request) << "MB"
+            << " / Total meomry pool size : " << get_mb_size(total_mem_pool_size) << "MB) ==========" << std::endl;
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/runtime/memory_pool.cpp:423]: typo: `meomry`

File: src/plugins/intel_gpu/src/runtime/debug_configuration.cpp [hunk 5]
```diff
@@ -149,6 +149,8 @@ static void print_help_messages() {
     message_list.emplace_back("OV_GPU_DisableRuntimeBufferFusing", "Disable runtime buffer fusing");
     message_list.emplace_back("OV_GPU_DisableMemoryReuse", "Disable memory reuse");
     message_list.emplace_back("OV_GPU_DumpRuntimeMemoryPool", "Dump memory pool contents of each iteration");
+    message_list.emplace_back("OV_GPU_DumpRuntimeMemoryPoolIters", "List of iteration's memory pool status");
+    message_list.emplace_back("OV_GPU_DumpRuntimeMemoryPoolPath", "Enable dumping memory pool status to csv file and set the dest path");
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/debug_configuration.cpp]: Done.

File: src/plugins/intel_gpu/src/graph/network.cpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/network.cpp]: Done.

File: src/plugins/intel_gpu/src/runtime/debug_configuration.cpp (hunk 2 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/debug_configuration.cpp]: Done

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp (hunk 4 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/memory_pool.cpp]: Done.

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp [hunk 6]
```diff
@@ -309,25 +319,146 @@ void memory_pool::clear_pool_for_network(uint32_t network_id) {
...
+        for (auto record : mem.second) {
+            if (record._type == type)
+                total_mem_pool_size += record._memory->size();
+        }
+    }
+    for (auto mem : _no_reusable_pool) {
+        if (mem.second._type == type)
+            total_mem_pool_size += mem.first;
+    }
+    return total_mem_pool_size;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/memory_pool.cpp:347]: How about to record the size when a new entry is allocated or released, instead of calculating every time?

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp [hunk 7]
```diff
@@ -309,25 +319,146 @@ void memory_pool::clear_pool_for_network(uint32_t network_id) {
...
+                            << "MB, type: " << record._type << ")'s usrs : " << std::endl;
+            float min_percentage = 100.0f;
+            float max_percentage = 0.f;
+            total_mem_pool_size += mem_size;
             for (auto user : record._users) {
-                GPU_DEBUG_COUT << "    --- " << user._prim_id << " (" << user._unique_id << std::endl;
+                float percentage = get_percentage(user._mem_size, mem_size);
+                min_percentage = std::min(percentage, min_percentage);
+                max_percentage = std::max(percentage, max_percentage);
+                total_mem_request += static_cast<float>(user._mem_size);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/memory_pool.cpp:434]: Seems that total mem request is calculated many times, I'dl like to recommend it to be stored.

File: src/plugins/intel_gpu/src/runtime/debug_configuration.cpp [hunk 8]
```diff
@@ -149,6 +149,8 @@ static void print_help_messages() {
     message_list.emplace_back("OV_GPU_DisableRuntimeBufferFusing", "Disable runtime buffer fusing");
     message_list.emplace_back("OV_GPU_DisableMemoryReuse", "Disable memory reuse");
     message_list.emplace_back("OV_GPU_DumpRuntimeMemoryPool", "Dump memory pool contents of each iteration");
+    message_list.emplace_back("OV_GPU_DumpRuntimeMemoryPoolIters", "List of iteration's memory pool status");
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/debug_configuration.cpp]: Done.

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp (hunk 6 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/memory_pool.cpp]: Done.

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp (hunk 7 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/memory_pool.cpp]: Done.

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp [hunk 9]
```diff
@@ -168,21 +220,30 @@ memory::ptr memory_pool::get_from_padded_pool(const layout& layout,
                 rec_list._memory->get_layout().format != format::fs_b_yx_fsv32 &&
                 layout.format != format::fs_b_yx_fsv32 &&
                 !has_conflict(rec_list._users, restrictions, network_id)) {
-                rec_list._users.insert({unique_id, network_id, prim_id});
+                rec_list._users.insert({unique_id, network_id, prim_id, layout.bytes_count()});
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/memory_pool.cpp:223]: Could we do this only when DumpMemoryPool is set? 
Beause calculating bytes_count can be another bottleneck too.

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp [hunk 10]
```diff
@@ -309,25 +412,152 @@ void memory_pool::clear_pool_for_network(uint32_t network_id) {
...
-    for (auto mem : _padded_pool) {
-        GPU_DEBUG_COUT << " layout: " << mem.first.to_short_string() << std::endl;
-        for (auto record : mem.second) {
-            GPU_DEBUG_COUT << "    " << record._memory->buffer_ptr() << ", type: " << record._type << ", users : " << std::endl;
-            for (auto user : record._users) {
-                GPU_DEBUG_COUT << "    --- " << user._prim_id << " (" << user._unique_id << std::endl;
+}
+
+void memory_pool::dump_to_screen(uint32_t net_id, uint32_t iter) {
+    GPU_DEBUG_COUT << "Dump memory pool of network (net_id : " << net_id << ", iter : " << iter << ")" << std::endl;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/memory_pool.cpp:475]: Could you fix these debug code only to be included when the DEBUG_CAP is ON?

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/memory_pool.cpp:534]: Done

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp (hunk 9 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/memory_pool.cpp]: Done.

File: src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp [hunk 11]
```diff
@@ -96,7 +99,7 @@ class memory_pool {
 
     std::multimap<uint64_t, memory_record> _non_padded_pool;
     std::map<layout, std::list<memory_record>, padded_pool_comparer> _padded_pool;
-    std::multimap<uint64_t, memory_record> _no_reusable_pool;
+    std::multimap<uint64_t, memory_record> _across_network_pool;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp:102]: Why did you change the name? Actually "across network reuse" is not considered to be used for now. (Deactivated) 
(In my thouhgt we even need to remove the legacy non used scenario of "reuse b/w network" to remove mislead ...) Changing name will cause more confusion.

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp [hunk 12]
```diff
@@ -50,24 +50,34 @@ bool memory_pool::has_conflict(const memory_set& a,
...
     {
-        auto it = _non_padded_pool.lower_bound(_layout.bytes_count());
+        auto it = _non_padded_pool.lower_bound(_layout_bytes_count);
 
         while (it != _non_padded_pool.end()) {
             if (it->second._network_id == network_id &&
                 it->second._type == type &&
                 it->second._memory->get_internal_params().mem == mem->get_internal_params().mem) {
-                auto user_it = it->second._users.find({ unique_id, network_id, prim_id });
+                auto user_it = it->second._users.find({ unique_id, network_id, prim_id, _layout_bytes_count});
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/memory_pool.cpp:66]: But still, I'd like to suggest to wrap this layout size with DEBUG_CODE to reduce unnecessary query element of the map.

File: src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp (hunk 11 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp]: Done.

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp (hunk 12 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/memory_pool.cpp]: Done

**Comment by ahnyoung-paul** [comment]
Removed ondnn cpu submodule from the commits. It has been added to the first commit by mistake during creating PR.

**Review by yeonbok — APPROVED** [review]

**Review by andrew-k-park — APPROVED** [review]
